src/storage/cache/queue.rs

```diff
@@ -0,0 +1,203 @@
+use core::marker::PhantomData;
+use std::{
+    ptr::{null_mut, NonNull},
+    sync::atomic::{AtomicPtr, Ordering::*},
+};
+
+pub struct Queue<T> {
+    size: usize,
+    head: AtomicPtr<Node<T>>,
+    tail: AtomicPtr<Node<T>>,
+}
+
+impl<T> Queue<T> {
+    pub fn new() -> Self {
+        let node = Node::new(None);
+        let node = Box::into_raw(Box::new(node));
+        Self {
+            head: AtomicPtr::new(node),
+            tail: AtomicPtr::new(node),
+            size: 0,
+        }
+    }
+
+    #[inline(always)]
+    pub fn push(&mut self, item: T) {
+        let node = Node::new(Some(item));
+        let node_ptr = Box::into_raw(Box::new(node));
+        let prev_tail = self.tail.load(Relaxed);
+        unsafe {
+            self.tail.swap(node_ptr, Release);
+            (*prev_tail).next.store(node_ptr, Release);
+        }
+
+        self.size += 1;
+    }
+
+    #[inline(always)]
+    pub fn pop(&mut self) -> Option<T> {
+        let mut head_ptr = NonNull::new(self.head.load(Relaxed)).unwrap();
+
+        loop {
+            match unsafe { head_ptr.as_mut().item.take() } {
+                Some(val) => {
+                    unsafe {
+                        let next = head_ptr.as_ref().next.load(Acquire);
+                        if !next.is_null() {
+                            self.head.swap(next, Relaxed);
+                        }
+                    };
+                    self.size -= 1;
+                    return Some(val);
+                }
```

Comment:
Memory leak, the `Node` object which contained the value taken after `unsafe { head_ptr.as_mut().item.take() }`  is still allocated.

Reply:
I need to fix this, thanks for catching this and mentioning it before too.

Reply:
Used std crate linkedlist as you'd suggested for this